Client Certificate Auth

jesshill99 opened the pull request.

jesshill99 committed:
Made initial list of possible certificate identification criteria.

jesshill99 committed:
Creating ClientCertificateConfiguration class and specs.

jesshill99 committed:
Implementing test for checking subject name of certificate

jesshill99 committed:
Moved certificate support into a new library

jesshill99 committed:
Continue to flesh out specs

jesshill99 committed:
Implement step to create a cert

jesshill99 committed:
Added steps to validate certificate is in store location. Added code to specs to dispose of test certificates.

jesshill99 committed:
Finalised ClientCertificateConfiguration specs so all are now passing

jesshill99 committed:
Started adding specs for cert auth.

jesshill99 committed:
Additional specs

jesshill99 committed:
Implemented validation of ClientIdWithCertificate

jesshill99 committed:
Added code that uses the authentication using a certificate code and added specs.

jesshill99 committed:
Implemented the spec for successfully getting a certificate based ClientIdentityConfiguration

jesshill99 committed:
Merge branch 'main' into feature/205-client-cert-auth

jesshill99 committed:
Update the $InvokeBuildModuleVersion parameter default value

jesshill99 committed:
Updated pipeline to use the 8.0 netSdkVersion

jesshill99 committed:
Added additionalNetSdkVersions 6.0

jesshill99 committed:
Update dependencies

jesshill99 committed:
Removed notes and added to an issue

jesshill99 committed:
Added package liicense

idg10 reviewed (commented):
We should probably bump the minor version, since this adds new functionality.

We should also add release notes.

jesshill99 committed:
Updated the package reference, the git version and the release notes.

HowardvanRooijen committed:
Update project configurations and dependencies

- Enhanced `pr-autoflow.json` to include new packages and remove wildcards.
- Modified `dependabot.yml` to exclude certain patterns for updates.
- Updated `auto_release.yml` with a new action version for improved retry logic.
- Introduced `build.yml` for a structured build process with concurrency settings.
- Updated `.gitignore` to ignore `.sbom.*` files in `local.settings.json`.
- Enhanced `config.ps1` with comments and examples for the build process.
- Updated `GitVersion.yml` to include `main` as a source branch and changed the next version to `4.0`.
- Changed target framework from `net6.0` to `net8.0` for several projects.
- Updated package references to newer versions for various libraries.
- Refactored initialization of `HashSet` and `List` in several files for modern syntax.
- Replaced NUnit assertions with `Assert.That` for better readability in tests.
- Refactored token provider classes for improved dependency injection handling.
- Extensively modified `packages.lock.json` for updated dependencies.
- Refactored test files for clarity and maintainability, removing obsolete code.

HowardvanRooijen committed:
Update dependencies and target framework for .NET 8.0

- Bumped `Endjin.RecommendedPractices.GitHub` to version `2.1.18` in both `Corvus.Identity.Abstractions.csproj` and `Corvus.Identity.Certificates.csproj`.
- Changed target frameworks in `Corvus.Identity.Certificates.csproj` to only support `.NET 8.0`.
- Removed obsolete package references and updated `Microsoft.Extensions.DependencyInjection.Abstractions` to a compatible version range.
- Added `Corvus.Identity.Certificates` as a project dependency in `packages.lock.json`.
- Updated `BouncyCastle.Cryptography` from `2.3.0` to `2.6.1` in `Corvus.Identity.Specs.csproj`.
- Removed obsolete SpecFlow code behind files from `Corvus.Identity.Specs.csproj`.
- Modified `LoadJsonConfiguration` method in `ConfigLoading.cs` to return a nullable type.

idg10 commented on Solutions/Corvus.Identity.Certificates/Corvus.Identity.Certificates.csproj:
```diff
@@ -19,15 +16,11 @@
 	    <PrivateAssets>all</PrivateAssets>
 	    <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
 	  </PackageReference>
+	  <PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="[8.0.*,)" />
```
These `[,)` style version specs turned out to be a mistake. I misunderstood what they did back when I introduced them, and I've been gradually removing them since.

At the time I thought these got written into the NuGet package. Turns out they don't: they get resolved at compile time to a specific version, meaning that the build is no longer deterministic. (If a new version comes out, you'll get that, meaning that the resulting NuGet package is different from what would have been built before that new version came out despite no change in source code.)

This form of package spec really only makes sense in applications (and test suites); in a library, it turns out to be no more useful than hard-coding it, with the added feature of making builds non-repeatable.

HowardvanRooijen committed:
Update Microsoft.Extensions package version formats

Changed version specifications for several Microsoft.Extensions packages from a range format (e.g., `[8.0.*, )`) to a wildcard format (e.g., `8.*`). This update was applied across multiple project files and the `packages.lock.json` file to allow for greater flexibility in package versioning.

HowardvanRooijen committed:
Hard code the v8.* package versions

idg10 commented on Documentation/ReleaseNotes/Corvus.Identity.v3.md:
```diff
@@ -1,5 +1,19 @@
 # Release notes for Corvus.Identity v3.
 
+## v3.4
```
I just realised that we're already post v4 aren't we? The GitVersion.yml has v4 in it. In which case this should be going into a new `Corvus.Identity.v4.md` file.

idg10 reviewed (approved):
v3 -> v4 release notes aside, this now LGTM

HowardvanRooijen committed:
Add release notes for Corvus.Identity v4.0 with client certificate authentication support and .NET 8.0 migration
